racket/collects/racket/file.rkt

```diff
@@ -134,100 +139,410 @@
+      "copy-from" copy-from))))
+
+(define (do-make-temporary-file/directory:bytes-append
+         ctxt prefix suffix copy-from base-dir
+         #:directory? directory?)
+  (define who
+    (if directory?
+        'make-temporary-directory*
+        'make-temporary-file*))
+  (check-bytes who prefix)
+  (check-bytes who suffix)
+  (check-base-dir who base-dir)
+  (unless (or directory?
+              (not copy-from)
+              (path-string? copy-from))
+    (raise-argument-error who
+                          "(or/c path-string? #f)"
+                          copy-from))
+  (define (make-name/bytes digits-str)
+    (bytes-append prefix
+                  ctxt
+                  (path-element->bytes (string->path-element digits-str))
+                  suffix))
+  (define (bad-result-msg details)
+    ;; i.e. the result is valid as path, but not for our purposes
+    (string-append "given prefix and suffix produced an invalid result;\n " details))
+  (do-make-temporary-file/directory:check-make-name
+   who copy-from base-dir
+   (λ (digits-str)
+     (bytes->path (make-name/bytes digits-str)))
+   #:wraped-make-name
+   (λ (digits-str)
+     (define bs (make-name/bytes digits-str))
+     (with-handlers ([exn:fail?
+                      (λ (e)
+                        (raise-arguments-error
+                         who
+                         "given prefix and suffix produced an invalid path"
+                         "prefix" prefix
+                         "suffix" suffix
+                         "produced" bs))])
+       (bytes->path bs)))
+   #:syntactic-directory-error
+   (λ (result)
+     (raise-arguments-error
+      who
+      (bad-result-msg "syntactic directory path not allowed")
+      "prefix" prefix
+      "suffix" suffix
+      "produced" result))
+   #:complete-with-base-error
+   (λ (result)
+     ;; On Windows, base-dir could be a drive specification,
+     ;; in which case it is ok for result to be an absolute path without a drive.
+     (raise-arguments-error
+      who
+      (bad-result-msg "complete path can not be combined with base-dir")
+      "prefix" prefix
+      "suffix" suffix
+      "produced" result
+      "base-dir" base-dir))))
+
+
+(define-for-syntax (syntax->tmp-context-string stx)
+  (define line   (syntax-line stx))
+  (define col    (syntax-column stx))
+  (define source (syntax-source stx))
+  (define pos    (syntax-position stx))
+  (define str-src
+    (cond [(path? source)
+           (regexp-replace #rx"^<(.*?)>(?=/)"
+                           (path->relative-string/library source)
+                           (lambda (_ s) (string-upcase s)))]
+          [(string? source) source]
+          [else #f]))
+  (define str-loc
+    (cond [(and line col) (format "-~a-~a" line col)]
+          [pos (format "--~a" pos)]
+          [else ""]))
+  (define combined-str (string-append (or str-src "rkttmp") str-loc))
+  (define sanitized-str (regexp-replace* #rx"[<>:\"/\\|]" combined-str "-"))
```

Comment:
It seems like this could have the same problem that came up in CI if the source file's path happened to contain a `~`: probably `sanatized-str` should handle that, too.